### DISCUSSION TITLE
how to populate a slice of compiled expr programs

### DISCUSSION THREAD

**Discussion by x1c1cle** [antonmedv/expr#742] [Q&A] [answered]
Hi All, 
I'd like to build a API driven expression engine that runs expression sets.
In order to do so, Id like to store compiled-programs in a slice, but I am unable to do so. 

here is my code with the issue:
[https://github.com/x1c1cle/expr-list-demo/blob/main/main.go](https://github.com/x1c1cle/expr-list-demo/blob/main/main.go)

I can store the compiled programs in an any slice, but I'm not able to assert to type *expr.Program or *vm.Program because neither are accessible/public to use. 

I'd appreciate any advice or help, Rhett

**Comment by antonmedv** [comment, marked as answer]
https://go.dev/play/p/FRHiy-r3RwI

```diff
package main

import (
	"fmt"
	"log"

	"github.com/expr-lang/expr"
+	"github.com/expr-lang/expr/vm"
)

func main() {
	fmt.Println("main start")

	env := map[string]any{
		"creditScore": 810,
	}

	expressions := []string{
		"creditScore >= 700",
		"creditScore >= 680",
		"creditScore >= 1000",
	}

	var programs []*vm.Program

	for _, exprStr := range expressions {
		program, err := expr.Compile(exprStr)
		if err != nil {
			log.Fatalf("failed to compile expression: %s", exprStr)
		}
		programs = append(programs, program)
	}

	for _, program := range programs {
		result, err := expr.Run(program, env)
		if err != nil {
			fmt.Println("error evaluating...")
		}
		fmt.Println("the result is: ", result)
	}
}
```

**Reply by x1c1cle** [reply]
perfect. thank you.
